src/Factory/RestControllerFactory.php

```diff
@@ -143,7 +143,6 @@ public function createServiceWithName(ServiceLocatorInterface $controllers, $nam
             ));
         }
 
-        $controller->setEventManager($events);
```

Comment:
It was bug and this commit fixed it. Please have a look on unit tests. As Matthew noted, the same instance was used for both - `Resource` and `RestController`.
Event `fetch` is actually a `Resource` event, not `RestController`.
Of course it could be BC break if listener was bound on wrong (bugged) event.